### PR TITLE
Avoid implicit os.fork() on Linux; improve comparison progress bar

### LINKED ISSUE #146: Linux Python 3.12 DeprecationWarning use of fork() may lead to deadlocks in the child.
## Summary:

Noticed locally on the cluster, and on recent Linux CI runs, e.g.

``DeprecationWarning: This process (pid=2079) is multi-threaded, use of fork() may lead to deadlocks in the child.``

This stems from cdf054c84c3cc86a7b2dd714d6b513d90bc34c8a to use ``rich.progress`` via ``multiprocessing`` added in PR #139.

Since Python 3.8 the default on macOS is "spawn", but currently the default on Linux is still "fork". The default on Linux will change in Python 3.14 to "forkserver".

## Description:

See also https://docs.python.org/3/library/os.html#os.fork and https://docs.python.org/3.12/library/multiprocessing.html#contexts-and-start-methods

## Reproducible Steps:

Please report a series of steps (e.g. command-line instructions) that would enable someone else to reproduce the issue. If possible, please attach a small set of input files/data that would be sufficient to allow someone else to reproduce the issue.

If it's not possible to reproduce the issue, please include a description of how you discovered the issue.

## Current Output:

Quoting the Linux Python 3.12 CI output:

```
=============================== warnings summary ===============================
tests/test_private_cli.py: 12 warnings
tests/test_public_cli.py: 4 warnings
  /home/runner/miniconda3/envs/pyani-github/lib/python3.12/multiprocessing/popen_fork.py:66: DeprecationWarning: This process (pid=2079) is multi-threaded, use of fork() may lead to deadlocks in the child.
    self.pid = os.fork()

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html

---------- coverage: platform linux, python 3.12.7-final-0 -----------
```

## Expected Output:

No DeprecationWarning about ``os.fork()``.

## pyani Version:

Noted with 2e2de18a6471e84b60e3c1d7c3668f8bc1029c73 but stems from cdf054c84c3cc86a7b2dd714d6b513d90bc34c8a to use ``rich.progress`` via ``multiprocessing`` added in PR #139.

## Python Version:

3.12.7

## Operating System:

Linux (macOS not affected)

### PR DESCRIPTION
**Pull request by peterjc**
This is to close #146 (avoid implicit ``os.fork()`` on Linux), but expanded to improve the progress-bar usage.

As of Python 3.8 onwards, the default on macOS ("Darwin") is "spawn"

As of Python 3.12, the default of "fork" on Linux triggers a deprecation warning. As of Python 3.14 on Linux the default will be "forkserver".

This changes our code to explicitly use "spawn" on macOS or "forkserver" otherwise.

For the progress bar, testing with SLURM on the Linux HPC revealed a problem with the initial thread setup. Originally I updated the progress bar on the main thread, with the single 'black-box' function call to snakemake on a child thread. That worked, but when snakemake failed we didn't get to see the traceback.

Furthermore, I fear the initial solution of monitoring the file system for the appearance of the target files will scale badly. This has now been changed to monitor progress via the database, and we can use the sqlite PRAGMA data_version to make this efficient.

## Type of change

- [x] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality not to work as expected)
- [ ] This change requires a documentation update
- [ ] This is a documentation update

## Action Checklist

- [x] Work on a single issue/concept (if there are multiple separate issues to address, please use a separate pull request for each)
- [ ] Fork the `pyani-plus` repository under your own account (please [allow write access for repository maintainers](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork))
- [x] Set up an appropriate development environment (please see `CONTRIBUTING.md` for this repository)
- [x] Create a new branch with a short, descriptive name
- [x] Work on this branch
  - [x] style guidelines have been followed (please see `CONTRIBUTING.md` for this repository)
  - [x] new code is commented, especially in hard-to-understand areas
  - [ ] corresponding changes to documentation have been made
  - [ ] tests for the change have been added that demonstrate the fix or feature works
- [x] Test locally with `pytest-plus -v` **non-passing code will not be merged**
- [x] Rebase against `origin/master`
- [x] Check changes with linting/formatting tools before submission (please see `CONTRIBUTING.md` for this repository)
- [x] Commit branch
- [x] Submit pull request, describing the content and intent of the pull request
- [x] Request a code review
- [x] Continue the discussion at [`Pull requests` section](https://github.com/widdowquinn/pyani-plus/pulls) in the `pyan-plus` repository

### CODE REVIEW

**Comment by peterjc** [comment]
Strange, not working. What did I break while finalising this... I was sure it did work!

**Comment by peterjc** [comment]
I've tried always using "spawn", but still the warning can be triggered depending how the tests are run (e.g. make test vs pytest, multiple workers or not). Unsure what the trigger is yet.

**Comment by peterjc** [comment]
I can't trigger this outside of pytest by calling our tool at the command line:

```
$ python --version
Python 3.12.7
$ rm -rf x.db && python -W always -m pyani_plus.public_cli fastani --database x.db --create-db --name 3V tests/fixtures/viral_example/
Indexing FASTAs ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:00:00 3/3
Database already has 0 of 3x3=9 comparisons, 9 needed
/users/xxx/repositories/pyani-plus/pyani_plus/public_cli.py:154: DeprecationWarning: About to use multiprocessing... might fork
  warnings.warn("About to use multiprocessing... might fork", DeprecationWarning)
Comparing pairs ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:00:02 9/9
Logged new run-id 1 for method fastANI with 3 genomes to database x.db
```

That was with a deliberate warning:

```
$ git diff
diff --git a/pyani_plus/public_cli.py b/pyani_plus/public_cli.py
index 98f80fa..0993c15 100644
--- a/pyani_plus/public_cli.py
+++ b/pyani_plus/public_cli.py
@@ -150,6 +150,8 @@ def run_snakemake_with_progress_bar(
     runner = snakemake_scheduler.SnakemakeRunner(workflow_name)
     # All rest replaces one line, runner.run_workflow(targets, params, workdir=working_directory)
 
+    import warnings
+    warnings.warn("About to use multiprocessing... might fork", DeprecationWarning)
     pending = [Path(_) for _ in targets]
     p = Process(
         target=runner.run_workflow,
```

This suggests the fork warning is triggered in connection with pytest and perhaps pytest-xdist?

File: pyani_plus/public_cli.py [hunk 1]
```diff
@@ -133,13 +135,48 @@ def check_fasta(fasta: Path) -> list[Path]:
...
+    run_id: int,
     targets: list[Path],
     params: dict,
     working_directory: Path,
     interval: float = 0.5,
-) -> int:
+) -> None:
     """Run snakemake with a progress bar.
 
     Currently this works by monitoring the filesystem for the appearance of
```
Comment (peterjc) [inline pyani_plus/public_cli.py:182]: I forgot to update the docstring...

File: pyani_plus/public_cli.py [hunk 2]
```diff
@@ -133,50 +135,82 @@ def check_fasta(fasta: Path) -> list[Path]:
...
+    done = 0
+    old_db_version = -1
+    with Progress(*progress_columns) as progress:
+        task = progress.add_task("Comparing pairs", total=total)
+        while done < total:
+            sleep(interval)
+            # Have there been any DB changes?
+            db_version = (
+                session.connection().execute(text("PRAGMA data_version;")).one()[0]
+            )
```
Comment (peterjc) [inline pyani_plus/public_cli.py:161]: Saw this fail on a cluster, ``sqlalchemy.exc.OperationalError: (sqlite3.OperationalError) disk I/O error``, might need a retry?

File: pyani_plus/public_cli.py [hunk 3]
```diff
@@ -133,50 +135,87 @@ def check_fasta(fasta: Path) -> list[Path]:
...
-            if p.exitcode is not None:
-                # Should be finished, but was it success or failure?
-                pending = []  # to break the loop
-    p.join()  # Should be immediate as should have finished
-    if p.exitcode is None:
-        msg = "Snakemake did not stop?"
-        sys.exit(msg)
-    return p.exitcode
+
+    # Call slurm!
```
Comment (widdowquinn) [inline pyani_plus/public_cli.py:205]: Is it always SLURM?

**Review by widdowquinn — APPROVED** [review]
This all looks good, but I'm not 100% sure how the caller would terminate the progress bar when necessary. Thanks @peterjc !

File: pyani_plus/public_cli.py (hunk 3 above)
Comment (peterjc) [reply pyani_plus/public_cli.py:205]: no, should be call snakemake - fixed on another PR

**Comment by peterjc** [comment]
We force it via ``p.terminate()`` in the main thread.

**Review by widdowquinn — APPROVED** [review]
Great - thanks @peterjc !

**Pull request merged**